esp-hal-common/src/rtc_cntl.rs

```diff
@@ -1,46 +1,608 @@
+            RtcSlowClock::RtcSlowClock8mD256 => HertzU32::Hz(17_500_000 / 256),
+        }
+    }
+}
+
+#[allow(unused)]
+#[derive(Debug, Clone, Copy)]
+/// Clock source to be calibrated using rtc_clk_cal function
+pub(crate) enum RtcCalSel {
+    /// Currently selected RTC SLOW_CLK
+    RtcCalRtcMux = 0,
+    /// Internal 8 MHz RC oscillator, divided by 256
+    RtcCal8mD256 = 1,
+    /// External 32 KHz XTAL
+    RtcCal32kXtal = 2,
+    #[cfg(not(feature = "esp32"))]
+    /// Internal 150 KHz RC oscillator
+    RtcCalInternalOsc = 3,
+}
+
+pub struct Rtc {
+    _inner: RTC_CNTL,
+    pub rwdt: Rwdt,
+    #[cfg(any(feature = "esp32c3", feature = "esp32s3"))]
+    pub swd: Swd,
+}
+
+impl Rtc {
     pub fn new(rtc_cntl: RTC_CNTL) -> Self {
-        Self { rtc_cntl }
+        rtc::init();
+        rtc::configure_clock();
+
+        Self {
+            _inner: rtc_cntl,
+            rwdt: Rwdt::default(),
+            #[cfg(any(feature = "esp32c3", feature = "esp32s3"))]
+            swd: Swd::new(),
+        }
+    }
+}
+
+/// RTC Watchdog Timer
+pub struct RtcClock;
+/// RTC Watchdog Timer driver
+impl RtcClock {
+    const CAL_FRACT: u32 = 19;
+
+    /// Get main XTAL frequency
+    /// This is the value stored in RTC register RTC_XTAL_FREQ_REG by the bootloader, as passed to
+    /// rtc_clk_init function.
+    fn get_xtal_freq() -> XtalClock {
+        let rtc_cntl = unsafe { &*RTC_CNTL::ptr() };
+        let xtal_freq_reg = rtc_cntl.store4.read().bits();
+
+        // Values of RTC_XTAL_FREQ_REG and RTC_APB_FREQ_REG are stored as two copies in
+        // lower and upper 16-bit halves. These are the routines to work with such a
+        // representation.
+        let clk_val_is_valid = |val| {
+            (val & 0xffffu32) == ((val >> 16u32) & 0xffffu32) && val != 0u32 && val != u32::MAX
+        };
+        let reg_val_to_clk_val = |val| val & u16::MAX as u32;
+
+        if !clk_val_is_valid(xtal_freq_reg) {
+            return XtalClock::RtcXtalFreq40M;
+        }
+
+        match reg_val_to_clk_val(xtal_freq_reg) {
+            40 => XtalClock::RtcXtalFreq40M,
+            #[cfg(any(feature = "esp32c3", feature = "esp32s3"))]
+            32 => XtalClock::RtcXtalFreq32M,
+            #[cfg(feature = "esp32")]
+            26 => XtalClock::RtcXtalFreq26M,
+            #[cfg(feature = "esp32")]
+            24 => XtalClock::RtcXtalFreq24M,
+            other => XtalClock::RtcXtalFreqOther(other),
+        }
+    }
+
+    /// Get the RTC_SLOW_CLK source
+    fn get_slow_freq() -> RtcSlowClock {
+        let rtc_cntl = unsafe { &*RTC_CNTL::ptr() };
+        let slow_freq = rtc_cntl.clk_conf.read().ana_clk_rtc_sel().bits();
+        match slow_freq {
+            0 => RtcSlowClock::RtcSlowClockRtc,
+            1 => RtcSlowClock::RtcSlowClock32kXtal,
+            2 => RtcSlowClock::RtcSlowClock8mD256,
+            _ => unreachable!(),
+        }
+    }
+
+    /// Select source for RTC_SLOW_CLK
+    fn set_slow_freq(slow_freq: RtcSlowClock) {
```

Comment:
this gives `warning: associated function is never used: `set_slow_freq`` when not targeting ESP32-C3 currently

Reply:
What would be the best solution for the scope of this PR? Should I restrict the implementation with `#[cfg(feature = "esp32c3")]` or is `#[allow(unused)]` fine?

Reply:
I guess `#[cfg(feature = "esp32c3")]` would be the cleaner way to get rid of this

Reply:
Actually, I just noticed that I can solve those warnings with proper code, the implementation of `rtc::configure_clock()` for the other chips is similar to ESP32-C3.
I'll submit a fix for them.

Reply:
Done!